### PR TITLE
Do not link static frameworks when inheriting search paths

### LINKED ISSUE #6065: Duplicate classes in tests from generated xcconfigs for test targets with target dependencies
🌈
- [x] I've read and understood the [CONTRIBUTING](https://github.com/CocoaPods/CocoaPods/blob/master/CONTRIBUTING.md) guidelines and have done my best effort to follow.
# Report
## What did you do?

I have a dynamic framework project with a handful of dependencies and a unit test target that lists the framework as a target dependency. My `Podfile` looks like this:

``` ruby
target 'FirebaseDatabaseUI' do
  use_frameworks!

  # Pods for Database
  pod 'Firebase/Database'

  target 'FirebaseDatabaseUITests' do
    inherit! :search_paths
  end
end
```

I noticed in one of the betas (definitely before 1.1.0.rc.2, I think around 1.1.0.beta.2) that duplicate classes from the dependencies were showing up in the unit tests runtime despite using `inherit! :search_paths`. After some hunting I found that CocoaPods was adding all of the test target's dependency's dependencies into the xcconfig file.

```
OTHER_LDFLAGS = $(inherited) -ObjC -l"c++" -l"icucore" -l"sqlite3" -l"z" 
  -framework "AddressBook" 
  -framework "CFNetwork" 
  -framework "CoreGraphics" 
  -framework "FirebaseAnalytics" 
  -framework "FirebaseCore" 
  -framework "FirebaseDatabase" 
  -framework "FirebaseInstanceID" 
  -framework "GoogleInterchangeUtilities" 
  -framework "GoogleSymbolUtilities" 
  -framework "GoogleUtilities" 
  -framework "Security" 
  -framework "StoreKit" 
  -framework "SystemConfiguration"
```

This issue is still present even if the tests target is moved out of its parent.

``` ruby
target 'FirebaseDatabaseUI' do
  use_frameworks!

  # Pods for Database
  pod 'Firebase/Database'
end

target 'FirebaseDatabaseUITests' do
  use_frameworks!
end
```

Deleting the `-framework` flags from the `xcconfig` file fixes the issue temporarily until the next `pod install`.

CocoaPods 1.0.1 doesn't have this issue, as it doesn't add any `ld` flags to the test target.

### Workaround

Deliberately not inheriting the OTHER_LDFLAGS (Other Linker Flags) build setting from CocoaPods' generated xcconfigs (i.e. delete `$(inherited)` in the project file) avoids this issue, but produces ugly warnings when running `pod install` and may cause issues in future versions of CocoaPods.

Removing the target dependency in Xcode, running `pod install`, and then re-adding the target dependency also works around the issue.
## What did you expect to happen?

I expected the test target to only inherit its parent's search paths and not include its parent's dependencies twice.
## What happened instead?

I had duplicate classes in the runtime.
### Stack

```
   CocoaPods : 1.1.1
        Ruby : ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-darwin15]
    RubyGems : 2.5.1
        Host : Mac OS X 10.11.6 (15G1004)
       Xcode : 8.0 (8A218a)
         Git : git version 2.10.0
Ruby lib dir : /Users/morganchen/.rbenv/versions/2.3.1/lib
Repositories : master - https://github.com/CocoaPods/Specs.git @ f0a7a1e1dde076ea332032c36e81d583e2b1fc6a
```
### Installation Source

```
Executable Path: /Users/morganchen/.rbenv/versions/2.3.1/bin/pod
```
### Plugins

```
cocoapods-deintegrate : 1.0.1
cocoapods-plugins     : 1.0.0
cocoapods-search      : 1.0.0
cocoapods-stats       : 1.0.0
cocoapods-trunk       : 1.1.1
cocoapods-try         : 1.1.0
```
### Podfile

``` ruby
target 'FirebaseDatabaseUI' do
  use_frameworks!

  # Pods for Database
  pod 'Firebase/Database'

  target 'FirebaseDatabaseUITests' do
    inherit! :search_paths
  end
end
```
## Project that demonstrates the issue

https://github.com/morganchen12/FirebaseUI-iOS/tree/pods-weird-behavior

Download this branch, run `pod install`, and then open `FirebaseUI.xcworkspace` and run tests in the `FirebaseDatabaseUI` scheme.

Thanks for your time!

### PR DESCRIPTION
**Pull request by dnkoutso**
closes #6065 

Tests to come and changelog

### CODE REVIEW

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 1]
```diff
@@ -95,12 +101,12 @@ def self.add_settings_for_file_accessors_of_target(target, xcconfig)
         #
         # @return [void]
         #
-        def self.add_static_dependency_build_settings(target, xcconfig, file_accessor)
+        def self.add_static_dependency_build_settings(aggregate_target, pod_target, xcconfig, file_accessor)
           file_accessor.vendored_static_frameworks.each do |vendored_static_framework|
-            XCConfigHelper.add_framework_build_settings(vendored_static_framework, xcconfig, target.sandbox.root)
+            XCConfigHelper.add_framework_build_settings(vendored_static_framework, xcconfig, pod_target.sandbox.root) unless aggregate_target != nil && aggregate_target.target_definition.inheritance == 'search_paths'
```
Comment (dnkoutso) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: This is the actual fix.

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! 
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* Do not link static frameworks when inheriting search paths 
  [dnkoutso](https://github.com/dnkoutso)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: spec/unit/generator/xcconfig/xcconfig_helper_spec.rb [hunk 2]
```diff
@@ -222,6 +222,38 @@ module XCConfig
...
+            xcconfig.to_hash['OTHER_LDFLAGS'].should == '-l"Bananalib"'
+          end
+
+          it 'should include static framework other ld flags when inheriting search paths' do
+            target_definition = stub(:inheritance => 'complete')
+            aggregate_target = stub(:target_definition => target_definition)
+            pod_target = stub(:sandbox => config.sandbox)
+            xcconfig = Xcodeproj::Config.new
+            @sut.add_static_dependency_build_settings(aggregate_target, pod_target, xcconfig, @accessor)
+            xcconfig.to_hash['OTHER_LDFLAGS'].should == '-l"Bananalib" -framework "Bananalib" -framework "StaticFramework"'
```
Comment (dnkoutso) [inline spec/unit/generator/xcconfig/xcconfig_helper_spec.rb]: This will require to update the test fixtures repo to include a fake StaticFramework I made.

Comment (dnkoutso) [inline spec/unit/generator/xcconfig/xcconfig_helper_spec.rb]: Unless we just use BananaLib.framework which has no executable in it and its being treated as a static framework

File: lib/cocoapods/generator/xcconfig/xcconfig_helper.rb [hunk 3]
```diff
@@ -95,12 +101,12 @@ def self.add_settings_for_file_accessors_of_target(target, xcconfig)
         #
         # @return [void]
         #
-        def self.add_static_dependency_build_settings(target, xcconfig, file_accessor)
+        def self.add_static_dependency_build_settings(aggregate_target, pod_target, xcconfig, file_accessor)
           file_accessor.vendored_static_frameworks.each do |vendored_static_framework|
-            XCConfigHelper.add_framework_build_settings(vendored_static_framework, xcconfig, target.sandbox.root)
+            XCConfigHelper.add_framework_build_settings(vendored_static_framework, xcconfig, pod_target.sandbox.root) unless !aggregate_target.nil? && aggregate_target.target_definition.inheritance == 'search_paths'
```
Comment (benasher44) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: nit: `unless !` reads a bit funny to me. could we just replace with `if` and reverse the 2nd condition?

Comment (dnkoutso) [inline lib/cocoapods/generator/xcconfig/xcconfig_helper.rb]: Done.

**Comment by Coeur** [comment]
This PR had regression consequences on #5514.

**Comment by jafara** [comment]
please do not revert this it fixes issue with reveal for me

**Comment by dnkoutso** [comment]
@jafara this will not be reverted but it will be enhanced to fix another issue that regressed.

**Comment by arjunaud** [comment]
I am also seeing this issue with static framework(FirebaseAnalytics). Could you please help me with this issue?

**Comment by Coeur** [comment]
@arjunaud please open an issue with detailed reproducible steps.

There is no need to comment on this 2.5 years old pull request unless you can precisely and with 100% confidence blame a specific line of code from the diff that introduced a regression without anyone else in the world noticing. That is, without anyone else reporting a problem that would have been in introduced by that pull request in CocoaPods 1.2.0. Or even without anyone working at Google complaining about integration issues with their highly popular framework.

**Pull request merged**